Feature/add event to anonymize data

lenonleite opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 7.x)
* a.b for any bug fixes (e.g. 5.2, 6.0)
* c.x for any bug fixes, features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 7.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌ <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | ✔️
| Deprecations?                          | ✔️
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️ <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

## Description

Feature responsible to create new event action that can anonymize data. 
It's based on issue #14119 



<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->
<!-- Remove HTML comment markup below to use the table for screenshots when relevant. -->
<!--
| Before                                 | After
| -------------------------------------- | ---
|                                        | 
-->


---
### 📋 Steps to test this PR:

<!--
This part is crucial. Take the time to write very clear, annotated and step by step test instructions, because testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Add new Contact with
    - Email
    - First Name
    - Last Name
    - Position
    - add tag 'Anonymize"
 
![mautic-issue-new-action-contact-form-1](https://github.com/user-attachments/assets/910b6f91-f530-4550-b6f8-f462ec628788)

3. Create a Segment name "_SegToAnonimyze_" with filter to add contacts with **Anonymize** tag
4. Run command `mautic:segments:update` to add contacts in segments
5. Create a Campaign based with segment SegToAnonimyze
6. Add new action with name "Anonymize User Data"
![Screenshot from 2024-10-07 14-08-16](https://github.com/user-attachments/assets/69e9fd33-6126-431d-8e6c-608a3b58e92e)

8. Fill form like:
    - Name : test
    - Execute this event : Immedialty
    - Pseudonymize data? Yes
    - Anonymize User Data from fields : Last Name
    - Delete User Data from fields : Position
    
![mautic-issue-new-action-form-edited](https://github.com/user-attachments/assets/30683fbc-14e1-4846-9b4f-199e1b49f59d)

8 - Save Campaign 
9 - Run command "mautic:campaigns:rebuild" to build campaign
10 - Run command  mautic:campaigns:trigger
11 - Come back to contact created and check:
    - Last name was change by hash
    - Position is empty

![mautic-issue-new-action-contact-form-edited](https://github.com/user-attachments/assets/f07e6f0f-3ade-498a-b808-05b20ddf3f03)

<!--
If you have any deprecations and backwards compatibility breaks, list them here along with the new alternative.
-->

lenonleite committed:
- add new event to Campaign events with the possibility of anonymizing data's.

lenonleite committed:
- add new event to Campaign events with the possibility of anonymizing data's.

lenonleite committed:
- fix phpstan
- fix fixcs

lenonleite committed:
- fix fixcs

shinde-rahul commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,506 @@
+{
+    public const KEY_EVENT_NAME = 'lead.action_anonymizeuserdata';
+
+    public const COLUMNS_ACEPPTED = ['text', 'longtext'];
```
```suggestion
    public const COLUMNS_ACCEPTED = ['text', 'longtext'];
```

lenonleite replied:
Done

shinde-rahul commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,506 @@
+<?php
+
+namespace Mautic\LeadBundle\EventListener;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\EventListener;
```
Missed strict type checking.

shinde-rahul commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,506 @@
+     *
+     * @return array<mixed>
+     */
+    private function setHashes(array $leadsCompanies, LeadField $field, bool $pseudonymize): array
```
Consider delegating this logic to a service class to improve the separation of concerns.

lenonleite replied:
done

shinde-rahul commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,506 @@
+    /**
+     * @param array<string, string|null> $field
+     */
+    private function setHash(
```
Consider delegating this logic to a service class to improve the separation of concerns.

lenonleite replied:
Done

shinde-rahul commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,506 @@
+        return $leadOrCompany;
+    }
+
+    private function formatHashEmail(string $email, int $limit): string
```
Consider delegating this logic to a service class to improve the separation of concerns.

lenonleite replied:
Done

shinde-rahul commented on app/bundles/LeadBundle/Form/Type/CampaignActionAnonymizeUserDataType.php:
```diff
@@ -0,0 +1,148 @@
+<?php
+
+namespace Mautic\LeadBundle\Form\Type;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\Form\Type;
```

shinde-rahul commented on app/bundles/LeadBundle/Form/Type/FieldListType.php:
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Mautic\LeadBundle\Form\Type;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\Form\Type;
```

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,50 @@
+<?php
+
+namespace Mautic\LeadBundle\Helper;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\Helper;
```

shinde-rahul commented on app/bundles/LeadBundle/Tests/EventListener/CampaignActionAnonymizeUserDataSubscriberTest.php:
```diff
@@ -0,0 +1,60 @@
+<?php
+
+namespace Mautic\LeadBundle\Tests\EventListener;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\Tests\EventListener;
```

shinde-rahul commented on app/bundles/LeadBundle/Tests/Form/Type/CampaignActionAnonymizeUserDataTypeTest.php:
```diff
@@ -0,0 +1,42 @@
+<?php
+
+namespace Mautic\LeadBundle\Tests\Form\Type;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\Tests\Form\Type;
```

shinde-rahul commented on app/bundles/LeadBundle/Tests/Functional/EventListener/CampaignActionAnonymizeUserDataSubscriberFunctionalTest.php:
```diff
@@ -0,0 +1,212 @@
+<?php
+
+namespace Mautic\LeadBundle\Tests\Functional\EventListener;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\Tests\Functional\EventListener;
```

shinde-rahul commented on app/bundles/LeadBundle/Tests/Functional/EventListener/CampaignActionAnonymizeUserDataSubscriberFormFunctionalTest.php:
```diff
@@ -0,0 +1,974 @@
+<?php
+
+namespace Mautic\LeadBundle\Tests\Functional\EventListener;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\Tests\Functional\EventListener;
```

shinde-rahul commented on app/bundles/LeadBundle/Tests/Helper/AnonymizeHelperTest.php:
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace Mautic\LeadBundle\Tests\Helper;
```
```suggestion

declare(strict_types=1);

namespace Mautic\LeadBundle\Tests\Helper;
```

shinde-rahul reviewed (changes requested):
Please add `declare(strict_types=1);` at the top of all new code files to ensure strict type enforcement. Also, suggest refactoring the listener according to SOLID principles by moving business logic to services and injecting dependencies via the constructor for improved testability and single responsibility.

lenonleite committed:
Apply suggestions from code review

declaring strict_types to each file

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Merge branch 'mautic:7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
Apply suggestions from code review

adding declare strict type

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Apply suggestions from code review

adding declare strict type

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
- add new service
- clean event listener

lenonleite committed:
- fixcs
- move config data of anonymize

lenonleite committed:
Merge remote-tracking branch 'origin/7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
- fix error wrong type propetires int to bool
- add tests

lenonleite committed:
- fix codeception

lenonleite committed:
- fix codeception
- fixcs, rector

lenonleite committed:
- fix codeception
- fixcs, rector

lenonleite committed:
- fix codeception
- fixcs, rector

lenonleite committed:
- fix codeception
- fixcs, rector

MadlenF reviewed (approved):
Works as described!

Bastian2718 reviewed (approved):
LGTM

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
- update test

shinde-rahul commented on app/bundles/LeadBundle/Form/Type/CampaignActionAnonymizeUserDataType.php:
```diff
@@ -0,0 +1,150 @@
+    {
+        return new Callback(
+            function ($validateMe, ExecutionContextInterface $context): void {
+                /** @var Integration $data */
```
```suggestion
```
`$data` is not an Integration entity.

shinde-rahul commented on app/bundles/LeadBundle/Form/Type/CampaignActionAnonymizeUserDataType.php:
```diff
@@ -0,0 +1,150 @@
+                ) {
+                    $context->buildViolation('mautic.lead.lead.events.error.empty_fields')
+                        ->addViolation();
+                    $data['properties']['fieldsToDelete']    = [];
+                    $data['properties']['fieldsToAnonymize'] = [];
```
Do you think this is needed? As during validation constraint, we should not alter the values.

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * @param string|bool|null $email
+     */
+    public static function email($email, bool $pseudonymized = false, string $newDomain = self::PRE_DEFINED_DOMAIN): string|false
+    {
+        if (empty($email)) {
+            return '';
+        }
+
+        if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+            return false;
+        }
```
```suggestion
    public static function email($email, bool $pseudonymized = false, string $newDomain = self::PRE_DEFINED_DOMAIN): ?string
    {
        if (empty($email) || !filter_var($email, FILTER_VALIDATE_EMAIL)) {
            return null;
        }
```

Lets avoid mixed return types (string, false, empty string).

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,52 @@
+        }
+        $emailParts = explode('@', $email);
+
+        $toEncrypt= $emailParts[0].time().random_int(0, 999999);
```
```suggestion
        $toEncrypt = $emailParts[0].time().random_int(0, 999999);
```

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,52 @@
+            $toEncrypt = $emailParts[0];
+            $newDomain = self::PRE_PSEUDONYMIZED_DOMAIN;
+        }
+        $name       = hash('sha256', $toEncrypt);
```
```suggestion

        $name = hash('sha256', $toEncrypt);
```

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * @param string|bool|null $email
+     */
+    public static function email($email, bool $pseudonymized = false, string $newDomain = self::PRE_DEFINED_DOMAIN): string|false
```
Can we rename the method name to something like `anonymizeEmail` for more verbosity?

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,52 @@
+    /**
+     * @param string|bool|null $text
+     */
+    public static function text($text, bool $pseudonymize = false): string
```
Can we rename the method name to something like `anonymizeText` for more verbosity?

shinde-rahul commented on app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php:
```diff
@@ -0,0 +1,378 @@
+    /**
+     * @param array<string, string|null> $field
+     */
+    private function setHash(
```
This block is a bit confusing due to the mix of data types and reassignments (e.g., $leadField being reused for both array and object). Consider adding a comment or splitting the logic for better readability.

shinde-rahul reviewed (changes requested):
Please add more test coverage. I’ve also suggested a few code-level improvements for clarity and maintainability.

Thanks @lenonleite.

kuzmany commented on app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php:
```diff
@@ -0,0 +1,378 @@
+    /**
+     * @var array<string, string>
+     */
+    public const COMPANY_FIELDS_TO_COLUMNS =
```
Don't understand this static list of fields. I assume we do not need it.

kuzmany commented on app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php:
```diff
@@ -0,0 +1,378 @@
+                continue;
+            }
+            $leadField = false;
+            if ($leadCompany instanceof Company && 'company' === $field->getObject()) {
+                $leadField = $leadCompany->getField($field->getAlias());
+                if (false === $leadField) {
+                    $alias          = self::COMPANY_FIELDS_TO_COLUMNS[$field->getAlias()] ?? $field->getAlias();
+                    $leadFieldValue = $leadCompany->getFieldValue($alias);
+                    if (property_exists($leadCompany, $alias) && null !== $leadFieldValue) {
+                        $tempLeadField = $leadCompany->{'get'.ucfirst($alias)}();
+                        if (empty($tempLeadField)) {
+                            continue;
+                        }
+                        unset($leadField);
+                        $leadField['value'] = $tempLeadField;
+                        $leadField['type']  = $field->getType();
+                    } else {
+                        continue;
+                    }
+                }
+            }
```
Do not see reason for this complexity of code. If we have company and alias we can use https://github.com/mautic/mautic/blob/7.x/app/bundles/LeadBundle/Model/CompanyModel.php#L235 for example

kuzmany commented on app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php:
```diff
@@ -0,0 +1,378 @@
+            if ('email' === $field['type']) {
+                $valueAnonymized = AnonymizeHelper::email($field['value'], $pseudonymize);
+                if ($leadField->getCharLengthLimit() < strlen($valueAnonymized)) {
+                    $valueAnonymized = $this->formatHashEmail($valueAnonymized, $leadField->getCharLengthLimit());
```
We can do this in one method, pass value, length and return and do not need so many if if cases

kuzmany reviewed (changes requested):
I like this feature. However, some parts of the code seem a bit difficult for me. I can only imagine a few lines instead of so many.

lenonleite committed:
- upgrade code quality.

lenonleite committed:
- remove upgrade after validating

lenonleite committed:
- remove upgrade after validating

lenonleite committed:
- upgrade tests

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

# Conflicts:
#	app/bundles/LeadBundle/Config/services.php

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
- improve code

lenonleite committed:
- remove static code

lenonleite committed:
Merge branch 'mautic:7.x' into 7.x

lenonleite committed:
- fix git

lenonleite committed:
Merge remote-tracking branch 'lenon/7.x' into 7.x

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
- improvement code
- split responsibilities
- add tests

lenonleite committed:
- add tests

lenonleite committed:
- fix phpstan
- fixcs

lenonleite committed:
- fix phpstan
- fixcs

lenonleite committed:
- fix phpstan
- fixcs

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

# Conflicts:
#	app/bundles/LeadBundle/Model/LeadModel.php

lenonleite committed:
- fix merge
- create audit log tests
- add anonymize contact company data test

lenonleite committed:
- fix merge
- create audit log tests
- add anonymize contact company data test
- add unit test to Submission model

lenonleite committed:
- add unit test to Submission model
- add test To Core Bundle

lenonleite committed:
- add unit tests

lenonleite committed:
- fix phpstan

lenonleite committed:
- improve sonar results

lenonleite committed:
- fixcs
- rector

lenonleite committed:
- fixcs
- rector

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
Merge remote-tracking branch 'lenon/feature/add-event-to-anonymize-data' into feature/add-event-to-anonymize-data

lenonleite committed:
- fixcs
- rector

lenonleite committed:
- fixcs
- rector

lenonleite committed:
- fixcs
- rector

lenonleite committed:
- fixcs
- phpstan

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite committed:
Merge branch '7.x' into feature/add-event-to-anonymize-data

lenonleite commented:
@adiux thank your for observation, I will work on it today, but can u give me bit more details to we don't have misunderstood... 🤗

kuzmany commented:
Also delete Contact Data in these fields — consider this a separate action or part of the Update Contact action. For the end user, it’s a hidden feature within another feature. If we have an anonymization feature, it should cover only anonymization and not other features.

Ideal scenario, user-friendly:

- Anonymize contact data (a simple, standalone action that performs the required changes).
- Update contact data action.
- Delete contact data (either as a separate action or included within Update contact data).

lenonleite committed:
- change messages

shinde-rahul commented on app/bundles/CoreBundle/Tests/Functional/Model/AuditLogModelTest.php:
```diff
@@ -0,0 +1,257 @@
+<?php
+
+// php
```
```suggestion
```

shinde-rahul commented on app/bundles/CoreBundle/Tests/Functional/Model/AuditLogModelTest.php:
```diff
@@ -0,0 +1,257 @@
+use Mautic\CoreBundle\Tests\Functional\CreateTestEntitiesTrait;
+use Symfony\Component\DependencyInjection\Container;
+
+class AuditLogModelTest extends MauticMysqlTestCase
```
```suggestion
final class AuditLogModelTest extends MauticMysqlTestCase
```

shinde-rahul commented on app/bundles/CoreBundle/Tests/Functional/Model/AuditLogModelTest.php:
```diff
@@ -0,0 +1,257 @@
+        $this->assertEmpty($logs, $message);
+    }
+
+    /** @param array<string,mixed> $criteria */
```
```suggestion
    /** 
     * @param array<string,mixed> $criteria 
     */
```

shinde-rahul commented on app/bundles/CoreBundle/Tests/Functional/Model/AuditLogModelTest.php:
```diff
@@ -0,0 +1,257 @@
+        $this->repo       = $this->model->getRepository();
+    }
+
+    // Helpers to reduce duplication
```
Should be removed. This is not making much sense.

lenonleite replied:
done

shinde-rahul commented on app/bundles/FormBundle/Model/SubmissionModel.php:
```diff
@@ -1194,4 +1197,71 @@ private function normalizeValue($value, Field $f): string
+        $columns         = $connection->createSchemaManager()->listTableColumns($nameTable);
+        $columnsToUpdate = [];
+        foreach ($columns as $column) {
+            // 1 = IntegerType
```
Should be removed.

lenonleite replied:
done

shinde-rahul commented on app/bundles/FormBundle/Tests/Model/SubmissionModelTest.php:
```diff
@@ -181,6 +182,11 @@ class SubmissionModelTest extends \PHPUnit\Framework\TestCase
     private \ReflectionClass $submissionModelReflection;
 
+    /**
+     * @var UserModel|(UserModel&object&MockObject)|(UserModel&MockObject)|(object&MockObject)|MockObject
```
```suggestion
     * @var UserModel|MockObject
```

shinde-rahul commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,144 @@
+use Mautic\LeadBundle\Services\AnonymizeContactCompanyData;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class CampaignActionAnonymizeUserDataSubscriber implements EventSubscriberInterface
```
```suggestion
final class CampaignActionAnonymizeUserDataSubscriber implements EventSubscriberInterface
```

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,88 @@
+
+        // if the limit is less than total characters keep going
+        // Extract the domain from the email
+        $atPosition = strrpos($email, '@'); // Find the position of '@'
```
```suggestion
        $atPosition = strrpos($email, '@');
```

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,88 @@
+            return $email;
+        }
+
+        // if the limit is less than total characters keep going
```
Did not get this.

shinde-rahul commented on app/bundles/LeadBundle/Helper/AnonymizeHelper.php:
```diff
@@ -0,0 +1,88 @@
+
+        // if the limit is less than total characters keep going
+        // Extract the domain from the email
+        $atPosition = strrpos($email, '@'); // Find the position of '@'
+
+        if (false === $atPosition) {
+            // If the email does not have a domain, return the email as-is
+            return $email;
+        }
+
+        $domain       = substr($email, $atPosition); // Extract the domain (e.g., @gmail.com or @uol.com)
```
Do we still need this? As you already are composing email address in ln#35 `$email      = $name.'@'.$newDomain;`

shinde-rahul commented on app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php:
```diff
@@ -0,0 +1,158 @@
+<?php
+
```
`declare(strict_types=1);` is missing.

shinde-rahul commented on app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php:
```diff
@@ -0,0 +1,158 @@
+use Mautic\LeadBundle\Model\FieldModel;
+use Psr\Log\LoggerInterface;
+
+class AnonymizeContactCompanyData
```
```suggestion
final class AnonymizeContactCompanyData
```
I have a suggestion regarding renaming the service, `ContactCompanyDataAnonymizer`.

shinde-rahul commented on app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    // To do, run tests here
```
```suggestion
```

shinde-rahul commented on app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php:
```diff
@@ -0,0 +1,158 @@
+        return $leadsCompanies;
+    }
+
+    // To do, run tests here
```
```suggestion
```

shinde-rahul commented on app/bundles/LeadBundle/Tests/EventListener/CampaignActionAnonymizeUserDataSubscriberTest.php:
```diff
@@ -0,0 +1,56 @@
+use Mautic\LeadBundle\Model\LeadModel;
+use PHPUnit\Framework\TestCase;
+
+class CampaignActionAnonymizeUserDataSubscriberTest extends TestCase
```
```suggestion
final class CampaignActionAnonymizeUserDataSubscriberTest extends TestCase
```

shinde-rahul commented on app/bundles/LeadBundle/Tests/Form/Type/CampaignActionAnonymizeUserDataTypeTest.php:
```diff
@@ -0,0 +1,242 @@
+use Mautic\LeadBundle\Model\FieldModel;
+use Symfony\Component\Form\FormBuilderInterface;
+
+class CampaignActionAnonymizeUserDataTypeTest extends \PHPUnit\Framework\TestCase
```
```suggestion
final class CampaignActionAnonymizeUserDataTypeTest extends \PHPUnit\Framework\TestCase
```

shinde-rahul commented on app/bundles/LeadBundle/Tests/Functional/EventListener/CampaignActionAnonymizeUserDataSubscriberFormFunctionalTest.php:
```diff
@@ -0,0 +1,1044 @@
+use Symfony\Component\DomCrawler\Crawler;
+use Symfony\Component\DomCrawler\Form;
+
+class CampaignActionAnonymizeUserDataSubscriberFormFunctionalTest extends MauticMysqlTestCase
```
```suggestion
final class CampaignActionAnonymizeUserDataSubscriberFormFunctionalTest extends MauticMysqlTestCase
```

shinde-rahul reviewed (changes requested):
Added some suggestions.

lenonleite committed:
Update app/bundles/CoreBundle/Tests/Functional/Model/AuditLogModelTest.php

remove comments

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Update app/bundles/CoreBundle/Tests/Functional/Model/AuditLogModelTest.php

change to final class

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Update app/bundles/CoreBundle/Tests/Functional/Model/AuditLogModelTest.php

improve comments

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Update app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php

remove comment

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Update app/bundles/LeadBundle/Services/AnonymizeContactCompanyData.php

remove comment

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Update app/bundles/LeadBundle/Tests/EventListener/CampaignActionAnonymizeUserDataSubscriberTest.php

change to final class

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Update app/bundles/LeadBundle/Tests/Form/Type/CampaignActionAnonymizeUserDataTypeTest.php

change to final class

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
Update app/bundles/LeadBundle/Tests/Functional/EventListener/CampaignActionAnonymizeUserDataSubscriberFormFunctionalTest.php

change to final class

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

lenonleite committed:
remove comments

ekkeguembel commented:
> @ekkeguembel Okay, let’s not complicate this.
> 
> 1. Campaign action: Anonymize Contact Data — clear to me now, thanks for update.
> 2. I recommend changing the Yes/No button for the Pseudonymization update to: Anonymization mode: Maximum / Basic (or something in this flow).
> 3. I recommend moving Delete Contact Data from these fields to a separate action (outside of Anonymization).
> 
> Points 2 and 3 are only recommendations and won’t block progress.
> 
> Decide and let me know, then finish the failing check and I will do the code review.

Thanks @kuzmany , much appreciated!

**Pseudonymization** should not be mistaken as "weak Anonymization", I think it's a fixed **technical term that we better not water down**, see [GDPR](https://www.edpb.europa.eu/system/files/2025-01/edpb_guidelines_202501_pseudonymisation_en.pdf) for instance.

I totally see your point for **separate action**, and we discussed the pros but also the cons: the list of actions is already long, "Delete Contact Data" is very similar to "Delete Contact", etc.
I'd prefer the renaming to "**"Reduce contact data (for privacy)"**" which is explicit about the goal, and covers all the features. If we feel that people don't get it, we can spin off another action later.

kuzmany commented:
Then let's keep it as Anonymize Contact Data , what is much clear. 

> Pseudonymization should not be mistaken as "weak Anonymization", I think it's a fixed technical term that we better not water down, see [GDPR](https://www.edpb.europa.eu/system/files/2025-01/edpb_guidelines_202501_pseudonymisation_en.pdf) for instance.

I understand everything, but imagine a 40-year-old, non-technical person who just wants to anonymize some data. What this user needs is  create a campaign, choose the Anonymize Contact Data action, and save. All the other options make them nervous.

But OK, I will do code review soon.
Thanks for patience

lenonleite committed:
- fixcs
 - Fix test after change messages

lenonleite committed:
- add tests to city and state

kuzmany commented on app/bundles/CoreBundle/Model/AuditLogModel.php:
```diff
@@ -70,4 +97,39 @@ public function getLogForObject($object, $id, $afterDate = null, $limit = 10, $b
+    /**
+     * @param array<int> $leadsId
+     */
+    public function deleteAuditLogByLeads(array $leadsId, bool $deleteCompaniesByLeads = false): void
```
What is goal of second parameter deleteAuditLogByLeads ?
Do not make sense, once method is deleteAuditLogByLeads then it should really do only what it is in the name

lenonleite replied:
Because in audit log we also have data as emails, name, last name that must be anonymize

kuzmany replied:
If we remove audit logs related to contacts, why do we need to run deleteCompaniesByLeads?
If I anonymize a first name, why do I need to remove companies' audit logs?
Why is there a deleteCompaniesByLeads parameter if the method is used in only one place?

kuzmany commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,144 @@
+                ['column' => 'l.id', 'expr' => 'in', 'value' => $event->getContactIds()],
+            ],
+        ];
+        $leads            = $this->leadModel->getEntities($leadFilter);
```
We do not need get entities, we already have in PendingEvent logs and of each log has lead entity.

kuzmany reviewed (commented):
For me this PR is over-engineered.
Lot of codes a lot changes, I spend hour for code review, I let some comments in the code,  but still not feel it.

Basically it works, but it should be written much simple.
This PR remove audit logs, modify form submission, modify companies (?). Nowhere mentioned, code is not readable, no logic. How form submissions are modified based on matched alias? Why we modify companies? 

I am not gonna block it,  but not gonna approve it. Community should take care and decide.

kuzmany commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,144 @@
+                'label'                  => 'mautic.lead.lead.events.anonymize',
+                'description'            => 'mautic.lead.lead.events.anonymize_descr',
+                // Kept for BC in case plugins are listening to the shared trigger
+                'eventName'              => LeadEvents::ON_CAMPAIGN_TRIGGER_ACTION, // @phpstan-ignore-line
```
I think this we do not need anymore

kuzmany commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,144 @@
+
+        foreach ($fields as $field) {
+            if (in_array($field->getId(), $properties['fieldsToDelete'])) {
+                [$leads,$companies]           = $this->setDeleteFields($leads, $companies, $field);
```
What the hell we doing with companies?

lenonleite replied:
We also anonimize fields from companies

kuzmany replied:
Thanks for not explaining. Again, if I choose to anonymize first and last names, why are companies being processed?

kuzmany commented on app/bundles/LeadBundle/EventListener/CampaignActionAnonymizeUserDataSubscriber.php:
```diff
@@ -0,0 +1,144 @@
+            }
+        }
+
+        if ($deleteFormResultsAndAuditLog) {
```
Does this variable mean if we do not have fields to anoynmize or remove we should stop processing campaign action?  Then we need variable  $deleteFormResultsAndAuditLog ?

lenonleite replied:
We anonymize data everywhere, clean data, forms, audit logs, leads, and companies. So yes, it istart from fields, if you have field to anonymize, check other places where can be writer.

lenonleite replied:
@kuzmany, it's not as simple as you think, on the contrary, it's very complex... the idea is simple, I wrote code twice because you didn't have the capacity to understand, and just drop random comments as this one. Still, this feature anonymizes all possible data that Mautic can generate, including forms, logs, leads, companies that following the Mautic flow, with the possibility of doing this differently, with random hashing, with pattern hashing, or simply deleting it.

Anyway @ekkeguembel decide with you all what u all want and I will com to close.

kuzmany replied:
>  I wrote code twice because you didn't have the capacity to understand, and just drop random comments as this one.

Please, can you avoid this? Random comments? WTF?
I understand you code, but it's  over-engineered.  You didn't do relevant answer to any of my question.

lenonleite replied:
First, "hell", "WTF"... 
If you have the time and a little humility, it will do you good to read and learn from this post. 
https://dev.to/tan/effective-comments-on-pull-requests-g46

Second, avoid this comment, ofc no, **who are u to ask it?** I didn't lie, as now you just say "over-engineered" or "I can only imagine a few lines instead of so many", so where? what is the best way? If you are lazy or don't like, just get out, these random comments don't help and just bring a bad environment to the community.

Third, I answered all of the questions before this unpleasant comment.

So let's be objective, if you wanna know about the goal of this PR, ask @ekkeguembel , auditlog, forms, the idea comes from him, if you have effective comments on the code, let's discuss, but I won't spend my time with, "over-engineered", or "I can only imagine a few lines instead of so many" and if you wanna work on it, feel free.

It is my last comment about it.

kuzmany replied:
Now I have a better understanding of the PR.  

The main issue for me is why, when I choose to anonymize a first name or reset its value, audit logs—including company audit logs and form submissions—are removed. This behavior isn’t explained, and the description doesn’t clarify it.

There are two possible solutions:

- Add a warning to the action. Anonymization affects every contact, just as we display a warning for contact deletion. Since the operation is irreversible, proceeding without a warning is risky.  
- Provide an option to retain logs and submissions instead of removing them.

My concerns have been clear, but they haven’t been addressed or reflected in the description or discussion. It’s difficult to continue the discussion if you feel offended.

lenonleite committed:
- save fields not mapped

lenonleite committed:
Merge remote-tracking branch 'lenon/feature/add-event-to-anonymize-data' into feature/add-event-to-anonymize-data

lenonleite committed:
- fix update field of different objects

lenonleite committed:
- fix phpstan

lenonleite committed:
- fix phpstan
